audit/hashstructure.go

```diff
@@ -207,20 +213,39 @@ func (w *hashWalker) Struct(v reflect.Value) error {
 		return nil
 	}
 
-	// If we aren't in a map value, return an error to prevent a panic
-	if v.Interface() != w.lastValue.Interface() {
-		return errors.New("time.Time value in a non map key cannot be hashed for audits")
+	if len(w.loc) < 2 {
+		// The last element of w.loc is reflectwalk.Struct, by definition.
+		// If len(w.loc) == 1 that means hashWalker.Walk was given a struct
+		// value and this is the very first step in the walk, and we don't
+		// currently support structs as inputs,
+		return errors.New("structs as direct inputs not supported")
 	}
 
-	// Create a string value of the time. IMPORTANT: this must never change
-	// across Vault versions or the hash value of equivalent time.Time will
-	// change.
-	strVal := v.Interface().(time.Time).Format(time.RFC3339Nano)
+	// Second to last element of w.loc is location that contains this struct.
+	switch w.loc[len(w.loc)-2] {
+	case reflectwalk.MapKey:
+		return errors.New("time.Time value in a map key cannot be hashed for audits")
```

Comment:
I assume the behavior generally is we only hash values, not keys. In which case throwing an error here seems overkill...why not just _not_ hash it but not throw an error and then when it's marshaled it'll become a string key value?

Reply:
> Suppose for example someday someone puts a birthdate (PII) in a map key, that could be a problem.

Sure -- but my understanding is if it had been preconverted to a string for the key, we'd let it through, because we only hash values. If I'm wrong and we hash keys already, then I guess we should hash it.

> Sure, I was trying to maintain the existing behaviour

So do we already throw errors for other types that appear in keys? I am assuming not since much of what comes in is a map[string]interface{}

Reply:
You're right, we don't hash keys already.  I'm perhaps being over-paranoid about failing to hash something we should.

> So do we already throw errors for other types that appear in keys? I am assuming not since much of what comes in is a map[string]interface{}

No, I simply meant that right now we already return an error if there's a time in a map key, so it seemed reasonable to maintain that.  I'm okay with letting it through unhashed instead, will try that.